### PR TITLE
Add custom error class and improve messaging for netifaces

### PR DESCRIPTION
**Pull request by zgoldman-r7**
Precursor to framework pr

Adds a custom error class to `ext/network_interface_ext/netifaces.c` to use on the ruby side of things for more specific errors. Also changes the description for the errors themselves to provide more detail.

Testing:
1) review the new error messages in the code to make sure they look good
2) ensure `bundle exec rake compile` runs without issues
3) to see the new error class:
  a) add `rb_raise(rb_cError, "Custom error");` to line 540 in `ext/network_interface_ext/netifaces.c`
  b) `bundle exec rake compile` 
  c) point `metasploit-framework` to the new gem locally
  d) in `metasploit-framework/lib/msf/core/opt_address_local.rb` make the following code changes:
```
13: def interfaces
14:   begin
15:     NetworkInterface.interfaces || []
16:   rescue => e
17:     binding.pry
18:     puts(e)
19:     elog(e)
20:     []
21:   end
22: end
```
  e) launch `./msfconsole`
  f) use the following commands in msfconsole:
```
msf6 > use windows/smb/psexec
[*] No payload configured, defaulting to windows/meterpreter/reverse_tcp
msf6 exploit(windows/smb/psexec) > set lhost localhost
lhost => localhost
msf6 exploit(windows/smb/psexec) > set rhost anything
rhost => whatever you put
msf6 exploit(windows/smb/psexec) > exploit
```
  g) when you hit the pry, check the value of `e`
  h) ensure the value is #<NetworkInterface::Error: Custom error>

### CODE REVIEW

**Comment by dwelch-r7** [comment]
@zgoldman-r7 if you rebase this PR it should hopefully pass on CI now

File: ext/network_interface_ext/netifaces.c [hunk 1]
```diff
@@ -200,6 +200,7 @@ static VALUE add_to_family(VALUE result, VALUE family, VALUE value)
   return result;
 }
 
+VALUE rb_cError;
```
[the remark this answers is not in the document]
Comment (zgoldman-r7) [reply ext/network_interface_ext/netifaces.c:203]: Reason for putting this here was for scoping as I was getting some undefined errors having it elsewhere, but I'll look into your suggestion and see if I can get it working!

[the remark this answers is not in the document]
Comment (adfoster-r7) [reply ext/network_interface_ext/netifaces.c]: Ah makes sense; forgot C has quirks here with forward references. I think you should be able to move both values to line 3

File: ext/network_interface_ext/netifaces.c [hunk 2]
```diff
@@ -200,6 +200,7 @@ static VALUE add_to_family(VALUE result, VALUE family, VALUE value)
   return result;
 }
 
+VALUE rb_cNetworkInterfaceError;
```
Comment (adfoster-r7) [inline ext/network_interface_ext/netifaces.c:203]: Wrote this in a separate comment, but just incase it gets lost:

I think you should be able to move both values to line 3 to fix the scope issue/introduce a bit of consistency

```c
#include "ruby.h"
#include "netifaces.h"

VALUE rb_cNetworkInterface
VALUE rb_cNetworkInterfaceError;

#if !defined(WIN32)
#if  !HAVE_GETNAMEINFO
#undef getnameinfo
#undef NI_NUMERICHOST
```

File: Gemfile.lock
Comment (dwelch-r7) [inline Gemfile.lock:1]: I think you need to revert this file

Comment (zgoldman-r7) [reply Gemfile.lock:1]: yep, good catch

**Comment by adfoster-r7** [comment]
@zgoldman-r7 Just needs a quick rebase on your side and we should be good to go 🤞

**Review by dwelch-r7 — APPROVED** [review]

**Pull request merged**